src/analysis/calculus/mean_value.lean

```diff
@@ -42,84 +42,177 @@ set_option class.instance_max_depth 120
 variables {E : Type*} [normed_group E] [normed_space ℝ E]
           {F : Type*} [normed_group F] [normed_space ℝ F]
 
-open metric set lattice asymptotics continuous_linear_map
+open metric set lattice asymptotics continuous_linear_map filter
+open_locale classical
+
+/-! ### Vector-valued functions `f : ℝ → E` -/
+
+section
+
+variables {f : ℝ → E} {a b : ℝ}
+
+/-- General fencing theorem for continuous functions with an estimate on the derivative.
+Let `f` and `B` be continuous functions on `[a, b]` such that
+
+* `∥f a∥ ≤ B a`;
+* `B` has a right derivative at every point `[a, b)`;
+* for each `x ∈ [a, b)` the right-side limit superior of the norm of `(f z - f x) / (z - x)`
+  is bounded above by a function `f'` which is strictly less than `B'`.
```

Comment:
The lemma is true with `limsup` replaced with `liminf`.

Reply:
How do I say `liminf f ≤ f'` with filters?

Reply:
You need cluster points. I think `liminf f L <= a` says that for every `b > a` there is a cluster point `c <= b` of the filter `map f L` (where "`c` clusters at `F`" means that for all `A \in F`, and `B \in nhds c`, `A  `A ∩ B != 0`).

Reply:
So,  `∀ b > a, ∀ s ∈ L.map f, (s ∩ Iio b).nonempty`?

Reply:
Or better `∀ b > a, ∀ s ∈ L, (s ∩ f ⁻¹' (Iio b)).nonempty`?

Reply:
Something like that, yes. You could also use the `liminf` defined in `src/order/liminf_limsup`, but I don't know how good the API is, and more importantly it only works well in a complete linear order, so you would need to work in `ereal`. While #1703 is not merged, this is not really an option.

Reply:
I proved with `∀ b > a, ∀ s ∈ L, (s ∩ f ⁻¹' (Iio b)).nonempty` (not pushed yet, because I made some unrelated modifications; splitting into PRs now).

Reply:
Then changed to `filter.frequently`.